hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1597,6 +1597,137 @@ private class Emit(
+        val rightBroadcastFlags = if (rNDims > 2) NDArrayEmitter.broadcastFlags(rightShapeArray) else Array[Code[Long]]()
+
+        val time = mb.newLocal[Long]
+
+        val lDataLength = mb.newField[Int]
+        val lDataLocation = mb.newField[Long]
+        val rDataLength = mb.newField[Int]
+        val rDataLocation = mb.newField[Long]
+
+        val setup = Code(
+          lT.setup,
+          rT.setup,
+          leftND := lT.value[Long],
+          rightND := rT.value[Long],
+          lDataLocation := lPType.data.load(region, leftND),
+          rDataLocation := rPType.data.load(region, rightND),
+          lDataLength := lPType.data.pType.loadLength(lDataLocation),
+          rDataLength := rPType.data.pType.loadLength(rDataLocation),
+          leftShapeArraySetup,
+          rightShapeArraySetup,
+          unifyShapeSetup
+//          Code.getStatic[java.lang.System, java.io.PrintStream]("out").invoke[Long, Unit](
+//            "println", time)
+        )
+
+        val outputPType = PNDArray(lPType.elementType, TNDArray.matMulNDims(lPType.nDims, rPType.nDims), true)
+
+        val elementMul = lPType.elementType match {
+          case PInt32(_) => (a: Code[_], b: Code[_]) => (coerce[Int](a) * coerce[Int](b)).asInstanceOf[Code[Any]]
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) * coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) * coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) * coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementAdd = lPType.elementType match {
+          case PInt32(_) => (a: Code[_], b: Code[_]) => (coerce[Int](a) + coerce[Int](b)).asInstanceOf[Code[Any]]
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) + coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) + coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) + coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementZero = lPType.elementType match {
+          case PInt32(_) => const(0).asInstanceOf[Code[Any]]
+          case PInt64(_) => const(0L).asInstanceOf[Code[Any]]
+          case PFloat32(_) => const(0.0f).asInstanceOf[Code[Any]]
+          case PFloat64(_) => const(0.0).asInstanceOf[Code[Any]]
+        }
+
+        val eVti = typeToTypeInfo(lPType.elementType.virtualType)
+
+        val emitter = new NDArrayEmitter(mb, outputPType.nDims, unifiedShapeArray, lPType.shape.pType, lPType.elementType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val seqIdxVars = idxVars.toSeq
+            val element = coerce[Any](mb.newField("matmul_element")(eVti))
+            val kLocal = mb.newLocal[Long]
+            val k = kLocal.load()
```

Comment:
This is because all of these arrays have to be of code long, and for some reason it doesn't seem to figure out the implicit conversion on things like `Array(kLocal)`

Reply:
Ah, seems to work if I do `Array[Code[Long]](kLocal)`. Will use that to address this comment so there's only `k`